File: tests/scriptTest.py
import pandas as pd

#Almacenar en un diccionario las posiciones. Ej: {'SMU': 123100, 'CENCOSUD': 312314, etc...}
def leer_posiciones(path_sf):
    sf_0406 = pd.read_excel(path_sf, engine='openpyxl')
    posiciones = sf_0406.set_index('CODIGO INSTRUMENTO')['POSICION DISPONIBLE'].to_dict()
    posiciones = {key.strip(): value for key, value in posiciones.items()}
    return posiciones

#Almacenar en un diccionario las suma total entre las compras y ventas de empresas, incluyendo todas las compras y ventas.
def calcular_operaciones_diarias(path_od):
    od_0306 = pd.read_excel(path_od, engine='xlrd')
    operaciones_diarias = {}
    for operacion, cantidad in zip(od_0306["Tipo Operación"], od_0306["Cantidad"]):
        if 'Venta' in operacion:
            nombre_venta = operacion[6:].upper()
            if nombre_venta in operaciones_diarias:
                operaciones_diarias[nombre_venta] -= cantidad
            else:
                operaciones_diarias[nombre_venta] = -cantidad
        else:
            nombre_compra = operacion[7:].upper()
            if nombre_compra in operaciones_diarias:
                operaciones_diarias[nombre_compra] += cantidad
            else:
                operaciones_diarias[nombre_compra] = cantidad
    return operaciones_diarias

#Hacer la suma entre los values del diccionario de servicios financieros y los values de operaciones diarias. 
#La suma entre S.F y O.D se almacena en un nuevo diccionario. 
def actualizar_posiciones(servicios_financieros, operaciones_diarias):
    resultado = servicios_financieros.copy()
    corta = {}
    simultanea = {}
    garantia = {}
    for key, value in operaciones_diarias.items():
        if key in resultado:
            resultado[key] += value
        else:
            tipo = input(f"{key} es simultanea, corta o garantia? ")
            if tipo == 'simultanea':
                if tipo in simultanea:
                    simultanea[key] += value
                else:
                    simultanea[key] = value
            elif tipo == 'corta':
                if tipo in corta:
                    corta[key] += value
                else:
                    corta[key] = value
            elif tipo == 'garantia':
                if tipo in garantia:
                    garantia[key] +=value
                else:
                    garantia[key] = value    
    return resultado, corta, simultanea, garantia

def comprobar_fechas(path_sf, path_od):
    od_0306 = pd.read_excel(path_od, engine='xlrd')
    sf_0406 = pd.read_excel(path_sf, engine='openpyxl')

    fecha_sf = sf_0406['FECHA'].iloc[0]
    fecha_od = od_0306['Fecha'].iloc[0]

    date_sf = pd.to_datetime(fecha_sf).normalize()
    date_od = pd.to_datetime(fecha_od).normalize()

    diferencia = date_sf - date_od
    if diferencia == 1:
        return True
    else: 
        return False

def main(path_sf, path_od):
    if comprobar_fechas(path_sf, path_od):
        print('Hola')
        return "No corresponden las fechas de los archivos subidos"
    else:    
        servicios_financieros = leer_posiciones(path_sf)
        print("POSICIONES SERVICIOS FINANCIEROS:")
        print(servicios_financieros)
        print('\n')

        operaciones_diarias = calcular_operaciones_diarias(path_od)
        print("NETO OPERACIONES DIARIAS:")
        print(operaciones_diarias)
        print("\n")

        resultado, corta, simultanea, garantia = actualizar_posiciones(servicios_financieros, operaciones_diarias)
        print('Resultado: ')
        print(resultado)
        print('\n')
        print("Corta:", corta)
        print("Simultanea:", simultanea)
        print("Garantia:", garantia)
        print('\n')



# Paths to the Excel files
path_sf = "./Excels/SERVICIOS FINANCIEROS 04:06.xlsx"
path_od = "./Excels/OPERACIONES DIARIAS 03:06.xls"

# Execute main function
main(path_sf, path_od)

File: tests/scriptTest.js
const xlsx = require('xlsx');
const prompt = require('prompt-sync')();

// Función para leer las posiciones
function leerPosiciones(path_sf) {
    const workbook = xlsx.readFile(path_sf);
    const sheet = workbook.Sheets[workbook.SheetNames[0]];
    const data = xlsx.utils.sheet_to_json(sheet, { header: 1 });
    const headers = data[0];
    const CODIGO_INSTRUMENTO_INDEX = headers.indexOf('CODIGO INSTRUMENTO');
    const POSICION_DISPONIBLE_INDEX = headers.indexOf('POSICION DISPONIBLE');

    const posiciones = {};
    for (let i = 1; i < data.length; i++) {
        const row = data[i];
        const codigoInstrumento = row[CODIGO_INSTRUMENTO_INDEX].trim();
        const posicionDisponible = row[POSICION_DISPONIBLE_INDEX];
        posiciones[codigoInstrumento] = posicionDisponible;
    }

    return posiciones;
}

// Función para calcular operaciones diarias
function calcularOperacionesDiarias(path_od) {
    const workbook = xlsx.readFile(path_od);
    const sheet = workbook.Sheets[workbook.SheetNames[0]];
    const data = xlsx.utils.sheet_to_json(sheet, { header: 1 });
    const headers = data[0];
    const TIPO_OPERACION_INDEX = headers.indexOf('Tipo Operación');
    const CANTIDAD_INDEX = headers.indexOf('Cantidad');

    const operacionesDiarias = {};
    for (let i = 1; i < data.length; i++) {
        const row = data[i];
        const operacion = row[TIPO_OPERACION_INDEX];
        const cantidad = row[CANTIDAD_INDEX];

        if (operacion.includes('Venta')) {
            const nombreVenta = operacion.substring(6).toUpperCase();
            if (operacionesDiarias[nombreVenta]) {
                operacionesDiarias[nombreVenta] -= cantidad;
            } else {
                operacionesDiarias[nombreVenta] = -cantidad;
            }
        } else {
            const nombreCompra = operacion.substring(7).toUpperCase();
            if (operacionesDiarias[nombreCompra]) {
                operacionesDiarias[nombreCompra] += cantidad;
            } else {
                operacionesDiarias[nombreCompra] = cantidad;
            }
        }
    }

    return operacionesDiarias;
}

// Función para actualizar posiciones
function actualizarPosiciones(serviciosFinancieros, operacionesDiarias) {
    const resultado = { ...serviciosFinancieros };
    const corta = {};
    const simultanea = {};
    const garantia = {};

    for (const [key, value] of Object.entries(operacionesDiarias)) {
        if (resultado[key] !== undefined) {
            resultado[key] += value;
        } else {
            const tipo = prompt(`${key} es simultanea, corta o garantia? `);
            if (tipo === 'simultanea') {
                simultanea[key] = (simultanea[key] || 0) + value;
            } else if (tipo === 'corta') {
                corta[key] = (corta[key] || 0) + value;
            } else if (tipo === 'garantia') {
                garantia[key] = (garantia[key] || 0) + value;
            }
        }
    }

    return { resultado, corta, simultanea, garantia };
}

// Función para comprobar las fechas
function comprobarFechas(path_sf, path_od) {
    const workbookSF = xlsx.readFile(path_sf);
    const sheetSF = workbookSF.Sheets[workbookSF.SheetNames[0]];
    const dataSF = xlsx.utils.sheet_to_json(sheetSF, { header: 1 });
    const fechaSF = dataSF[1][dataSF[0].indexOf('FECHA')];

    const workbookOD = xlsx.readFile(path_od);
    const sheetOD = workbookOD.Sheets[workbookOD.SheetNames[0]];
    const dataOD = xlsx.utils.sheet_to_json(sheetOD, { header: 1 });
    const fechaOD = dataOD[1][dataOD[0].indexOf('Fecha')];

    const dateSF = new Date(fechaSF).setHours(0, 0, 0, 0);
    const dateOD = new Date(fechaOD).setHours(0, 0, 0, 0);

    return (dateSF - dateOD) === 86400000; // 86400000 ms = 1 day
}

// Función principal
function main(path_sf, path_od) {
    if (comprobarFechas(path_sf, path_od)) {
        console.log('No corresponden las fechas de los archivos subidos');
        return;
    }

    const serviciosFinancieros = leerPosiciones(path_sf);
    console.log('POSICIONES SERVICIOS FINANCIEROS:');
    console.log(serviciosFinancieros);
    console.log('\n');

    const operacionesDiarias = calcularOperacionesDiarias(path_od);
    console.log('NETO OPERACIONES DIARIAS:');
    console.log(operacionesDiarias);
    console.log('\n');

    const { resultado, corta, simultanea, garantia } = actualizarPosiciones(serviciosFinancieros, operacionesDiarias);
    console.log('Resultado: ');
    console.log(resultado);
    console.log('\n');
    console.log('Corta:', corta);
    console.log('Simultanea:', simultanea);
    console.log('Garantia:', garantia);
    console.log('\n');
}

// Paths to the Excel files
const path_sf = "./Excels/SERVICIOS FINANCIEROS 04:06.xlsx";
const path_od = "./Excels/OPERACIONES DIARIAS 03:06.xls";

// Execute main function
main(path_sf, path_od);